solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1643,6 +1660,83 @@ public ScoreMode scoreMode() {
     qr.setDocList(new DocSlice(0, sliceLen, ids, scores, totalHits, maxScore));
   }
 
+  CollectorManagerResult searchCollectorManagers(int len, QueryCommand cmd, Query query,
+      boolean needTopDocs, boolean needMaxScore, boolean needDocSet) throws IOException {
+    CollectorManager<MultiCollector, CollectorManagerResult> manager = new CollectorManager<MultiCollector, CollectorManagerResult>() {
+      @Override
+      public MultiCollector newCollector() throws IOException {
+        // nocommit: Here, creating a MultiCollector for every segment (correctness > speed).
+        // Need to explore sharing a single MultiCollector with every segment. Are these
+        // sub-collectors thread-safe? DocSetCollector seems like not thread-safe, does someone know?
```

Comment:
DocSetCollector is *not* threadsafe.  That is something Etsy tried to tackle.  I think if we had per-segment DocSets (and thus segment oriented filter cache), then this would be a non-issue?  I don't think you should tackle this in this issue; it's a TODO for another day.